eng/fv/refinement.rst

```diff
@@ -0,0 +1,554 @@
+
+Each SPIN output file ``mymodel-I.spn`` 
+is converted to a C test runner file ``tr-mymodel-I.c``
+by concatenating the following components:
+
+``mymodel-pre.h``
+    This is a fairly standard first part of an RTEMS test C file.
+    It is used unchanged.
+
+refined test segments
+  The annotations in ``mymodel-I.spn`` are converted, in order,
+  into test code snippets using the refinement file ``mymodel-rfn.yml``.
+  Snippets are gathered into distinct code segments based on the Promela
+  process number reported in each annotation.
+  Each code segment is used to construct a C function called
+  ``TestSegmentP()``, where ``P`` is the relevant process number.
+
+``mymodel-post.h``
+    This is static code that declares the top-level RTEMS Tasks
+    used in the test.
+    These are where the code segments above get invoked.
+
+``mymodel-run.h``
+   This defines top-level C functions that implement a given test runner. The top-level function has a name like ``RtemsMyModel_Run``
+   This is not valid C as it needs to produce a name parameterized by 
+   the relevant scenario number. It contains Python string format substitution 
+   placeholders that allow the relevant number to be added to end of the 
+   function name. So the above run function actually appears in this file as ``RtemsMyModel_Run{0}``, so ``I`` will be substituted in for ``{0}`` to result in the name ``RtemsMyModel_RunI``.
+   In particular, it invokes ``TestSegment0()`` which contains code 
+   generated from Promela process 0, which is the main model function.
+   This declares test variables, and initializes them.
+
+These will generate test-runner test files as follows:
+
+.. code-block:: none
+
+  tr-mymodel-0.c
+  ...
+  tr-mymodel-{N-1}.c
+
+In addition, the test case file ``tc-mymodel.c`` needs to have entries added
+manually of the form below, for ``I`` in the range 0 to N-1.:
+
+.. code-block:: c
+
+  T_TEST_CASE( RtemsMyModelI )
+  {
+    RtemsMyModel_RunI(
+      ...
+    );
+  }
+
+These define the individual test cases in the model, each corresponding to precisely one SPIN scenario.
+
+Test Code Deployment
+^^^^^^^^^^^^^^^^^^^^
+
+All files starting with ``tc-`` or ``tr-`` are copied to the
+relevant testsuite directory.
+At present, this is ``testsuites/validation`` at the top level in
+the ``rtems`` repository.
+All the names of the above files with a ``.c`` extension are added
+into a YAML file that
+defines the Promela generated-test sources.
+At present, this
+is ``spec/build/testsuites/validation/model-0.yml``
+at the top-level in the ``rtems`` repository.
+They appear in the YAML file under the ``source`` key:
+
+.. code-block:: yaml
+
+  source:
+  - testsuites/validation/tc-mymodel.c
+  - testsuites/validation/tr-mymodel-0.c
+  ...
+  - testsuites/validation/tr-mymodel-{N-1}.c
+  - testsuites/validation/ts-model-0.c
+
+Performing Tests
+^^^^^^^^^^^^^^^^
+
+At this point, ``waf`` can be invoked in the main ``rtems`` directory as normal,
```

Comment:
might be better to write generically, "At this point build rtems as normal, e.g., with `waf` -- just a minor way to avoid baking too much rigidity in this document